### PR TITLE
[Instrumentation.AspNet] Update semantic conventions for metrics

### PR DESCRIPTION
**Pull request by qhris**
Updates metrics to match [semantic conventions for 1.23](https://github.com/open-telemetry/semantic-conventions/blob/v1.23.0/docs/http/http-metrics.md#http-server).

This was previously discussed in #1407.

## Changes

Replaces `http.server.duration` with the new `http.request.server.duration` metric and adds missing attributes.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes

## Remaining issues

* [x] ~~Unable to get `Application_Error` to fire in my sample application in order to verify `error.type` for exceptions. It works fine for status codes.~~ Excluded for now.
* [x] Unable to write a test for `network.protocol.version` which uses `request.ServerVariables["SERVER_PROTOCOL"]` because that property/class is heavily locked down. I have verified this to work on a sample application.
* [x] The `HttpRequestRouteHelper` works fine for fetching the route template but it is marginally useful because the route easily ends up like `/{controller}/{action}` which doesn't say much.

### CODE REVIEW

**Comment by qhris** [comment]
Hey,

I'm leaving this up as a draft until I can finish up the remaining tasks and run some more tests.
I could however use some input from @Kielek and @vishweshbankwar for the direction of this PR and the remaining tasks?

**Comment by vishweshbankwar** [comment]
> Hey,
> 
> I'm leaving this up as a draft until I can finish up the remaining tasks and run some more tests. I could however use some input from @Kielek and @vishweshbankwar for the direction of this PR and the remaining tasks?

@qhris 
1) Let's handle `error.type` outside of this PR. It would be tricky to get that value on metrics side. Take a look at https://github.com/open-telemetry/opentelemetry-dotnet/pull/4986 if you can do something similar here.
2) I see the test creates `HttpContext.Current`. Is it possible to do  `HttpContext.Current.Request.ServerVariables.Add` and validate that you get the `network.protocol.version`? 
3) For `http.route`, this is a known issue. Please open a separate issue for this if there isn't one out there already. For now, you can copy how tracing is doing it.

**Comment by qhris** [comment]
@vishweshbankwar thanks for the input!

1. Thanks for the link, I can look into this a bit more. Are you suggesting to remove it entirely for this PR if not?
2. A tried a variety of things to get it working.
    - `ServerVariables.Add()` throw with `System.NotSupportedException`.
    - `ServerVariables` only has a getter so I can't assign to it.
    - I also tried setting the backing field `_serverVariables` through reflection but ended up running into an issue since the type `HttpServerVarsCollection` is internal to `System.Web`. I could try to instantiate that type through reflection as well but I didn't feel comfortable with that as it would get very messy. Might be the only way to go though.
    - I tried making `HttpRequest` a mock but that throws: `System.NotSupportedException : Type to mock (HttpRequest) must be an interface, a delegate, or a non-sealed, non-static class`.
    - It seems using `HttpContextBase` is the recommended way to make the code more testable but that would require making quite a bigger change to the codebase (i.e. rewriting `HttpTelemetryModule`, `ActivityHelper`, et al.).
3. Sounds good 👍

**Comment by vishweshbankwar** [comment]
> @vishweshbankwar thanks for the input!
> 
> 1. Thanks for the link, I can look into this a bit more. Are you suggesting to remove it entirely for this PR if not?
> 2. A tried a variety of things to get it working.
>    
>    * `ServerVariables.Add()` throw with `System.NotSupportedException`.
>    * `ServerVariables` only has a getter so I can't assign to it.
>    * I also tried setting the backing field `_serverVariables` through reflection but ended up running into an issue since the type `HttpServerVarsCollection` is internal to `System.Web`. I could try to instantiate that type through reflection as well but I didn't feel comfortable with that as it would get very messy. Might be the only way to go though.
>    * I tried making `HttpRequest` a mock but that throws: `System.NotSupportedException : Type to mock (HttpRequest) must be an interface, a delegate, or a non-sealed, non-static class`.
>    * It seems using `HttpContextBase` is the recommended way to make the code more testable but that would require making quite a bigger change to the codebase (i.e. rewriting `HttpTelemetryModule`, `ActivityHelper`, et al.).
> 3. Sounds good 👍

Yes - A separate PR for `error.type`.

For `network.protocol.version`, You could remove this one too for now in favor of making progress. I will try to take a look at it when I get chance.

**Comment by qhris** [comment]
Hey, I'm made some additional findings here.

The reason why I wasn't able to get `Application_Error` to fire was because I was testing a `ASP Web API` application. This type of app has a completely different error flow as desribed in [this Microsoft documentation](<https://learn.microsoft.com/en-us/aspnet/web-api/overview/error-handling/web-api-global-error-handling>). 

This means that traces do not behave as expected either, since the error status of the span will never be set [here](<https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInListener.cs#L184>).

For now, I went ahead and removed the `error.type` as suggested.

Would it be a blocker to leave in the protocol version as is? It is a very useful metric and it would be a shame to leave it out.

**Comment by vishweshbankwar** [comment]
> Would it be a blocker to leave in the protocol version as is

No, I think it is ok.

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs:67]: Remove `network.protocol.name` -- it is conditionally required if not defaults to `http`.
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#metric-httpserverrequestduration

Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs]: Thanks, I missed that 👍

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs:66]: Needs special handling for httpMethod. see https://github.com/open-telemetry/opentelemetry-dotnet/pull/5001

File: src/Shared/SemanticConventions.cs
Comment (vishweshbankwar) [inline src/Shared/SemanticConventions.cs:120]: remove this.

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpRequestRouteHelper.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpRequestRouteHelper.cs:26]:
```suggestion
internal sealed class HttpRequestRouteHelper
```

File: test/OpenTelemetry.Instrumentation.AspNet.Tests/HttpInMetricsListenerTests.cs
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNet.Tests/HttpInMetricsListenerTests.cs:148]:
```suggestion
        // Unable to check for "network.protocol.version" because we can't set server variables due to the accessibility
```

File: src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md:11]: Mention that this is a **breaking change**

**Comment by Kielek** [comment]
@qhris, I think that. semantic convention also for spans/traces has to be adjusted. It can be done in separate PR, but it should be done before the release.

**Comment by carlos-neto-maersk** [comment]
Can you guys please check https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1431 ?  
  
It is related to this, because due to the changes done to telemetry module to have the activity stopped on this `OnActivityStoppedCallback`, under load we are experiencing that weird pattern (and I think that it is because it because it is closing activities that aren't to be closed in other `OnActivityStoppedCallback` handlers.  
  
I'm working on not closing the activity after the handlers, and using the current timestamp to figure it out the duration instead of using `Activity.Duration`, but the code isn't great.  
  

**And thank you so much for your great work.**

**Comment by qhris** [comment]
@carlos-neto-maersk I have seen the issue but I haven't had the time to check it out yet. I'll keep the discussion to that issue/PR as it is not exactly related to this one.

**Comment by qhris** [comment]
> @qhris, I think that. semantic convention also for spans/traces has to be adjusted. It can be done in separate PR, but it should be done before the release to make the transition from unstable to stable world only once.

@Kielek It should go in a different PR agreed. This one should probably be worked out before though.

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs
Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs]: I went ahead and implemented something similar, while also handling `OTEL_INSTRUMENTATION_HTTP_KNOWN_METHODS` properly.

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs:72]: nit: Take a look at this PR to refactor requestMethodHelper.
https://github.com/open-telemetry/opentelemetry-dotnet/pull/5026

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs:26]: why not use dictionary?

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs
Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInMetricsListener.cs:72]: Yeah that looks cleaner, I'll rewrite it as that.

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs
Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs:26]: Since we're on a hot code path I figured looping over a small set of items should be faster than using a dictionary, but I can change this back if necessary.

Since you asked I decided to do a small performance test and found this:
image

I can change it to a dictionary to match the other implementation you linked.

**Comment by qhris** [comment]
This is a getting a bit long at this point, but I noticed that `SemanticConventions.cs` now has a lot of deprecated and unused properties such as `AttributeNetTransport`. Should I remove the relevant ones or leave that for future cleanup?

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs:55]: Make this static? You could make the whole class static.

Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs:55]: That would not be possible in this case.
The way it is written it allows the unit tests to re-instatiate the class which will re-read the environment variables in order to test `OTEL_INSTRUMENTATION_HTTP_KNOWN_METHODS`.

Making it all static will make it untestable.

I think it is fine in this case because the class will only be instantiated twice (once in `HttpInMetricsListener` and in the future `HttpInListener`.

**Comment by Kielek** [comment]
> This is a getting a bit long at this point, but I noticed that `SemanticConventions.cs` now has a lot of deprecated and unused properties such as `AttributeNetTransport`. Should I remove the relevant ones or leave that for future cleanup?

As it is in Shared folder, my vote is for separate PR.

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs:32]: nit: instantiate in the ctor itself based on the value of env variable?

**Review by vishweshbankwar — APPROVED** [review]
LGTM: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1429/files#r1396542874 should be addressed before merging.

**Comment by vishweshbankwar** [comment]
@qhris Could you please open an issue for following up on network.protocol.version testing?

File: src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs
Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/Implementation/RequestMethodHelper.cs:32]: Yeah that makes sense.

File: src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md
Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md]: I'm so glad you posted this, I just realized I didn't check this properly when testing.

The buckets have not been updated, the implementation [here](<https://github.com/samlam/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/Metric.cs#L32>) does not actually take `OpenTelemetry.Instrumentation.AspNet` into consideration.

I updated the `CHANGELOG`, fixed it, and added tests for it locally now. But if we want it fixed upstream this PR will have to wait.
I also had to pull in `OpenTelemetry` as a dependency as that's the only way to access `AddView`.

**Comment by qhris** [comment]
Please do not merge this until [this comment](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1429#discussion_r1397501852) is resolved.

**Comment by qhris** [comment]
> @qhris Could you please open an issue for following up on network.protocol.version testing?

@vishweshbankwar I created an issue for making things more testable here: #1438.

File: src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj:27]: this is really against OpenTelemetry principles, and should not be done.
Please do either of the below:
1. Just documenting this in readme and ask users to add views themselves.
2. Request to make OTel sdk give special treatment to this one also. (I am supportive of this idea).

**Review by cijothomas — CHANGES_REQUESTED** [review]
See https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1429/files#r1397902272

Requesting to change the dependency on SDK before merge.

File: src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md]: @qhris Good catch. We should not be adding SDK adding dependency to instrumentation as Cijo mentioned. Could you remove that? and send a PR upstream to include ASPNET in the list?

Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md]: I made the suggested changes, and I also updated the buckets to remove the zero bucket as was done in https://github.com/open-telemetry/opentelemetry-dotnet/pull/5021.
I posted a link to the upstream PR in response to Cijo.

File: src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj
Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj]: Yes, I figured that was the case, I just wanted to make the tests pass earlier, I completely agree it's not a good idea.

I removed it and went with option 2 for now, I opened at issue at https://github.com/open-telemetry/opentelemetry-dotnet/pull/5063 for now.

I have removed the dependency again but leaving the tests to fail until we decide whether to wait on the upstream PR.

File: src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md:17]: Need to explicitly mention that this buckets are coming because OTel Sdk is hardwired to specially treat this metrics, and mention the starting version of OTel SDK as well. (once we know the version)

Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md:17]: Sounds good to me, but I'll hold off until we know.

**Review by Kielek — APPROVED** [review]
@cijothomas, can we merge it with current wording?

**Comment by cijothomas** [comment]
> @cijothomas, can we merge it with current wording?

Since we know the exact version of the SDK which is required to get correct histogram buckets, I'd suggest to call it out in the changelog.
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/CHANGELOG.md#170-rc1

File: src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md
Comment (qhris) [reply src/OpenTelemetry.Instrumentation.AspNet/CHANGELOG.md:17]: I've went ahead with the suggested changes.

**Comment by qhris** [comment]
Fixed all the remaining issues but the tests won't pass until the SDK is updated.

**Comment by utpilla** [comment]
@qhris Thanks for your patience!

**Pull request merged**
